Installation issue: adios2

<!-- Thanks for taking the time to report this build failure. To proceed with the report please:

1. Title the issue "Installation issue: <name-of-the-package>".
2. Provide the information required below.

We encourage you to try, as much as possible, to reduce your problem to the minimal example that still reproduces the issue. That would help us a lot in fixing it quickly and effectively! -->

### Steps to reproduce the issue

<!-- Fill in the exact spec you are trying to build and the relevant part of the error message -->
```console
$ spack install adios2 +endian_reverse +python
```

### Information on your system

<!-- Please include the output of `spack debug report` -->
```
$ spack debug report
* **Spack:** 0.16.1-2651-fc392d2f56
* **Python:** 3.8.5
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** original
```

<!-- If you have any relevant configuration detail (custom `packages.yaml` or `modules.yaml`, etc.) you can add that here as well. -->

### Additional information

<!-- Please upload the following files. They should be present in the stage directory of the failing build. Also upload any config.log or similar file if one exists. -->
* [spack-build-out.txt](https://github.com/spack/spack/files/6481152/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/6481155/spack-build-env.txt)

<!-- Some packages have maintainers who have volunteered to debug build failures. Run `spack maintainers <name-of-the-package>` and @mention them here if they exist. -->

@ax3l  @chuckatkins  @williamfgc

### General information

<!-- These boxes can be checked by replacing [ ] with [x] or by clicking them after submitting the issue. -->
- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

Might be a good upstream report, cc @pnorbert
```
adios2/helper/adiosMemory.cpp:32:22: error: no match for 'operator==' (operand types are 'const adios2::DataType' and 'const char [1]')
   32 |         if (destType == "")
      |             ~~~~~~~~ ^~ ~~
      |             |           |
      |             |           const char [1]
      |             const adios2::DataType
```
Upstream: https://github.com/ornladios/ADIOS2 -> posted as https://github.com/ornladios/ADIOS2/issues/2700

AFAIK, `endian_reverse` is one of the less-intense tested features.

--- a/var/spack/repos/builtin/packages/adios2/2.7.1-fix-broken-endian-reverse-compile.patch
+++ b/var/spack/repos/builtin/packages/adios2/2.7.1-fix-broken-endian-reverse-compile.patch
@@ -1,0 +1,23 @@
+From 5aed0d5c3ecb7afbd56a786243b2bc8811e0fbdd Mon Sep 17 00:00:00 2001
+From: Axel Huebl <axel.huebl@plasma.ninja>
+Date: Fri, 14 May 2021 18:22:19 -0700
+Subject: [PATCH] Fix Broken Endinage Reverse Compile
+
+Add right type in comparison.
+---
+ source/adios2/helper/adiosMemory.cpp | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/source/adios2/helper/adiosMemory.cpp b/source/adios2/helper/adiosMemory.cpp
+index c3a4246a91..3bf8c89adc 100644
+--- a/source/adios2/helper/adiosMemory.cpp
++++ b/source/adios2/helper/adiosMemory.cpp
+@@ -29,7 +29,7 @@ void CopyPayloadStride(const char *src, const size_t payloadStride, char *dest,
+ #ifdef ADIOS2_HAVE_ENDIAN_REVERSE
+     if (endianReverse)
+     {
+-        if (destType == "")
++        if (destType == DataType::None)
+         {
+         }
+ #define declare_type(T)                                                        \

--- a/var/spack/repos/builtin/packages/adios2/package.py
+++ b/var/spack/repos/builtin/packages/adios2/package.py
@@ -212,6 +212,10 @@ class Adios2(CMakePackage, CudaPackage, ROCmPackage):
         sha256="8221073d1b2f8944395a88a5d60a15c7370646b62f5fc6309867bbb6a8c2096c",
     )
 
+    # Fix mismatched datatypes in comparison
+    # See https://github.com/ornladios/ADIOS2/pull/2701
+    patch("2.7.1-fix-broken-endian-reverse-compile.patch", when="@2.7.1")
+
     # cmake: find threads package first
     # https://github.com/ornladios/ADIOS2/pull/3893
     patch("2.9.2-cmake-find-threads-package-first.patch", when="@2.9.2:")

